### PR TITLE
Global seed foundation

### PR DESCRIPTION
**Pull request by RichSweetWafer**
Hi!

I've recently encountered a problem during training process of our model. The problem itself is not so important (division by zero during back propagation in cross_entropy_error.cpp), because I can see that `dev` branch has changed a lot since the last release version. But I see that what I lacked during debugging has not been addressed yet: reproducibility. Randomization is good, but controlled randomization is even better. Frameworks like PyTorch try to provide ways to reduce the number of nondeterministic behavior. One way is to implement a global RNG, which is what was done in this PR.

The user just has to write something like
```
std::random_device rd;
setGlobalSeed(rd());
```
before using the library.

This is a somewhat "foundational" code, some basic stuff to improve upon in the long run. Now, I don't know the codebase as well as you guys do, so I expect some notes being made and mistakes highlighted, if you will consider this idea.

### CODE REVIEW

File: opennn/layer.h [hunk 1]
```diff
@@ -245,14 +246,24 @@ class Layer
     template <int Rank>
     void dropout(Tensor<type, Rank>& tensor, const type& dropout_rate) const
     {
+        if (dropout_rate >= type(1))
+        {
+            throw runtime_error("Dropout rate should be in range [0, 1). Got: " + std::to_string(dropout_rate));
```
Comment (RichSweetWafer) [inline opennn/layer.h:251]: Also, there probably should be a runtime error if dropout function uses rate that is too high

**Comment by RubyAM** [comment]
Hi! 

Thank you for reaching out and for your detailed explanation. We completely agree that reproducibility and controlled randomization are absolutely essential when debugging training issues, which is why we have actually just addressed this in the latest version of the library! As you can see in our recent updates to `random_utilities.cpp` and `random_utilities.h`, we have already implemented a global seed mechanism very similar to what you are suggesting.

Users can now easily control the deterministic behavior of the library by calling:

```cpp
opennn::set_seed(your_seed_value);
```

To ensure reproducibility even during parallel execution, our implementation uses thread_local generators. If a global seed is set, each OpenMP thread initializes its own generator deterministically based on the global seed and its thread ID (global_seed + thread_id * 5489u).

We highly recommend pulling the latest changes to take advantage of this and many other recent updates. Let us know if you run into any other issues or have further feedback once you try it out.

Thanks again!
